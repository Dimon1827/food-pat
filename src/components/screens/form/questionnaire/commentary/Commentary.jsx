import { useFormContext } from 'react-hook-form';
import { useState } from 'react';
import './Commentary.scss';

function Commentary() {
  const {
    register,
    formState: { errors },
  } = useFormContext();
  const [remainLetters, setRemainLetters] = useState('');
  const numberLetters = 200;
  const remainNumbersLetters = numberLetters - remainLetters.length;

  return (
    <div className="container">
      <div className="questionnaire__field commentary">
        <h2 className="commentary__title form-title">Комментарий</h2>
        <textarea
          className="commentary__text"
          {...register('comment', {
            maxLength: {
              value: numberLetters,
              message: 'Сообщение превысило кол-во допустимых букв',
            },
          })}
          value={remainLetters}
          onChange={(e) => setRemainLetters(e.target.value)}
          cols="30"
          rows="10"
          placeholder="Расскажите о вас. Что предпочитаете и т.д"
        ></textarea>
        {remainNumbersLetters >= 0 ? (
          <span className="commentary__counter">
            Осталось букв: {remainNumbersLetters}
          </span>
        ) : (
          errors.comment && (
            <div className="information__error">{errors.comment.message}</div>
          )
        )}
      </div>
    </div>
  );
}

export default Commentary;
