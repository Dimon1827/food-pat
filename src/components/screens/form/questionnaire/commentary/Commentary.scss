@import 'src/assets/styles/variables';

.commentary__title {
  margin-bottom: 38px;

  @media (min-width: $tablet-width) {
    margin-bottom: 84px;
  }

  @media (min-width: $desktop-width) {
    margin-bottom: 77px;
  }
}

.commentary__text {
  font-family: $oswald-font;
  padding: 11px 14px 0 15px;
  width: 100%;
  height: 150px;
  font-size: 16px;
  text-transform: uppercase;
  border: 1px solid $special-grey-3;
  margin-bottom: 18px;


  @media (min-width: $tablet-width) {
    height: 175px;
    margin-bottom: 59px;
    padding: 9px 14px 0 20px;

    &::placeholder {
      font-size: 20px;
      line-height: 30px;
    }
  }

  @media (min-width: $desktop-width) {
    margin-bottom: 65px;
  }
}
