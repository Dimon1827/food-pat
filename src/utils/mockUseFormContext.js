import { useFormContext } from "react-hook-form";

export function mockUseFormContext(errors) {
  useFormContext.mockReturnValue({
    register: jest.fn(),
    formState: {
      errors: errors,
    },
  });
}


