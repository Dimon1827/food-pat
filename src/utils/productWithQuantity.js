const productWithQuantity = (food) => {
  return {
    ...food,
    isChecked: false,
    quantity: 1,
  };
};

export default productWithQuantity;
