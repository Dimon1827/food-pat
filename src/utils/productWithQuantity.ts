import { IProducts } from "redux/slice/productsSlice.types";

interface IproductWithQuantity extends Omit<IProducts, 'isChecked' | 'quantity'> {};

const productWithQuantity = (food: IproductWithQuantity) => {
  return {
    ...food,
    isChecked: false,
    quantity: 1,
  };
};

export default productWithQuantity;
