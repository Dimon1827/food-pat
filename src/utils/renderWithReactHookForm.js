import { render } from '@testing-library/react';
import { FormProvider, useForm } from 'react-hook-form';

export function renderWithReactHookForm(ui, { defaultValues = {} } = {}) {
  let reactHookFormMethods = {};

  const Wrapper = ({ children }) => {
    const methods = useForm({ defaultValues });
    return <FormProvider {...methods}>{children}</FormProvider>;
  };

  return {
    ...render(ui, { wrapper: Wrapper }),
  };
}