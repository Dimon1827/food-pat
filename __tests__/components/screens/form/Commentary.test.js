import { screen } from '@testing-library/react';
import { renderWithReactHookForm } from '@/utils/renderWithReactHookForm';
import Commentary from '@/components/screens/form/questionnaire/commentary/Commentary';
import userEvent from '@testing-library/user-event';
import React from 'react';

jest.mock('react-hook-form', () => ({
  ...jest.requireActual('react-hook-form'),
  useFormContext: () => ({
    register: jest.fn(),
    formState: {
      errors: {
        comment: {
          message: 'Сообщение превысило кол-во допустимых букв',
        },
      },
    },
  }),
}));

describe('Commentary', () => {
  it('Commentary renders correctly', () => {
    renderWithReactHookForm(<Commentary />);
    expect(screen.getByTestId('commentary')).toMatchSnapshot();
  });
  it('The number of letters must be counted correctly', async () => {
    renderWithReactHookForm(<Commentary />);
    const placeholder = 'Расскажите о вас. Что предпочитаете и т.д';
    const text = 'Привет';
    const textarea = screen.getByPlaceholderText(placeholder);
    await userEvent.type(textarea, text);
    expect(textarea).toHaveValue(text);
    const counter = screen.getByText('Осталось букв: 194');
    expect(counter).toBeInTheDocument();
  });
  it('should display an error message when the maximum length is exceeded', async () => {
    const setRemainLetters = jest.fn();
    const placeholder = 'Расскажите о вас. Что предпочитаете и т.д';
    const text = 'b'.repeat(300);
    jest
      .spyOn(React, 'useState')
      .mockImplementation(() => [text, setRemainLetters]);

    
    const errorMessage = 'Сообщение превысило кол-во допустимых букв';
    
    renderWithReactHookForm(<Commentary />);
    const textarea = screen.getByPlaceholderText(placeholder);
    await userEvent.type(textarea, text);
    const errorElement = await screen.findByText(errorMessage);
    expect(errorElement).toBeInTheDocument();
  });
});
