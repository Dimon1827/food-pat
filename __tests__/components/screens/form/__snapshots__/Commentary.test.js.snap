// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Commentary Commentary renders correctly 1`] = `
<div
  class="questionnaire__field commentary"
  data-testid="commentary"
>
  <h2
    class="commentary__title form-title"
  >
    Комментарий
  </h2>
  <textarea
    class="commentary__text"
    cols="30"
    placeholder="Расскажите о вас. Что предпочитаете и т.д"
    rows="10"
  />
  <span
    class="commentary__counter"
  >
    Осталось букв: 
    200
  </span>
</div>
`;
